File: 19th/코딩테스트 심화반/week4/1744번_수 묶기.cpp
#include <iostream>
#include <queue>

using namespace std;

priority_queue<int, vector<int>, greater<int>> n_pq; // 0보다 작거나 같은 정수는 오름차순 저장
priority_queue<int> p_pq; // 0보다 큰 정수는 내림차순 저장

int main() {
    int N;
    cin >> N;

    int num;
    for (int i = 0; i < N; i++) {
        cin >> num;
        if (num <= 0) n_pq.push(num);
        else p_pq.push(num);
    }

    int ans = 0;

    // top에 위치한 두 원소를 곱하여 더하면 가장 큰 값이 됨
    // 만약 top에 한 원소만 있을 경우 그대로 더함
    int first, second; 
    while (!n_pq.empty()) {
        first = n_pq.top();
        n_pq.pop();
        if (n_pq.empty()) {
            ans += first;
            break;
        }
        second = n_pq.top();
        n_pq.pop();

        ans += first * second;
    }

    // top에 위치한 두 원소를 곱하여 더하면 가장 큰 값이 됨
    // 만약 top에 한 원소만 있을 경우 그대로 더함
    // 1의 경우 곱해도 크기가 커지지 않기에, 그대로 더함
    while (!p_pq.empty()) {
        first = p_pq.top();
        p_pq.pop();
        if (p_pq.empty()) {
            ans += first;
            break;
        }
        second = p_pq.top();
        p_pq.pop();

        if (first != 1 && second != 1) ans += first * second;
        else ans += first + second;
    }

    cout << ans;
}

File: 19th/코딩테스트 심화반/week4/1744번_수 묶기.py
n = int(input())

n_arr = []       # 음수를 저장할 리스트
p_arr = []       # 양수를 저장할 리스트
zero_cnt = 0     # 0의 개수를 세기 위한 변수

# 숫자들을 분류
for _ in range(n):
    i = int(input())
    if i < 0:
        n_arr.append(i)  # 음수
    elif i == 0:
        zero_cnt += 1    # 0
    else:
        p_arr.append(i)  # 양수

n_arr.sort()  # 음수는 오름차순 (작은 수부터)
p_arr.sort()  # 양수도 오름차순 (큰 수부터 pop 하기 위함)

ans = 0  # 최종 결과

# 음수 처리
while len(n_arr) > 1:
    first = n_arr.pop(0)   # 가장 작은 음수
    second = n_arr.pop(0)  # 그 다음 작은 음수
    ans += first * second  # 두 음수를 곱하면 양수가 되므로 이득

# 남은 음수가 하나 있을 경우
if n_arr:
    if not zero_cnt:       # 0이 없으면 그냥 더함
        ans += n_arr[0]
    # 0이 있으면 곱해서 없애는 것과 같으므로 아무 것도 안 함

# 양수 처리
while p_arr:
    first = p_arr.pop()  # 가장 큰 수부터

    if not p_arr:
        ans += first  # 짝이 안 맞으면 그냥 더함
        break

    second = p_arr.pop()

    # 둘 중 하나라도 1이면 더하는 게 이득
    if first != 1 and second != 1:
        ans += first * second
    else:
        ans += first + second

print(ans)  # 최종 결과 출력